src/rgw/rgw_client_io.h

```diff
@@ -4,83 +4,237 @@
+    return get_decoratee().complete_header();
+  }
+
+  size_t recv_body(char* const buf, const size_t max) override {
+    return get_decoratee().recv_body(buf, max);
+  }
+
+  size_t send_body(const char* const buf,
+                   const size_t len) override {
+    return get_decoratee().send_body(buf, len);
+  }
+
+  void flush() override {
+    return get_decoratee().flush();
+  }
+
+  RGWEnv& get_env() noexcept override {
+    return get_decoratee().get_env();
+  }
+
+  size_t complete_request() override {
+    return get_decoratee().complete_request();
+  }
+};
+
+} /* namespace rgw */
+} /* namespace io */
+
+
+/* We're doing this nasty thing only because of extensive usage of templates
+ * to implement the static decorator pattern. C++ templates de facto enforce
+ * mixing interfaces with implementation. Additionally, those classes derive
+ * from RGWRestfulIO defined here. I believe that including in the middle of
+ * file is still better than polluting it directly. */
+#include "rgw_client_io_decoimpl.h"
+
+
+/* RGWRestfulIO: high level interface to interact with RESTful clients. What
+ * differentiates it from rgw::io::RestfulClient is providing more specific APIs
+ * like rgw::io::Accounter or the AWS Auth v4 stuff implemented by filters
+ * while hiding the pipelined architecture from clients.
+ *
+ * rgw::io::Accounter came in as a part of rgw::io::AccountingFilter. */
+class RGWRestfulIO : public rgw::io::AccountingFilter<rgw::io::RestfulClient*> {
+  SHA256 *sha256_hash;
+
+public:
+  virtual ~RGWRestfulIO() {}
+
+  RGWRestfulIO(rgw::io::RestfulClient* engine)
+    : AccountingFilter<rgw::io::RestfulClient*>(std::move(engine)),
+      sha256_hash(nullptr) {
+  }
+
+  using DecoratedRestfulClient<RestfulClient*>::recv_body;
+  virtual int recv_body(char* buf, size_t max, bool calculate_hash);
+  std::string grab_aws4_sha256_hash();
+}; /* RGWRestfulIO */
 
-  int print(const char *format, ...);
-  int write(const char *buf, int len);
-  virtual void flush() = 0;
-  int read(char *buf, int max, int *actual, bool hash = false);
 
-  string grab_aws4_sha256_hash();
+/* Type conversions to work around lack of req_state type hierarchy matching
+ * (e.g.) REST backends (may be replaced w/dynamic typed req_state). */
+static inline rgw::io::RestfulClient* RESTFUL_IO(struct req_state* s) {
+  return static_cast<rgw::io::RestfulClient*>(s->cio);
+}
 
-  virtual int send_status(int status, const char *status_name) = 0;
-  virtual int send_100_continue() = 0;
-  virtual int complete_header() = 0;
-  virtual int send_content_length(uint64_t len) = 0;
+static inline rgw::io::Accounter* ACCOUNTING_IO(struct req_state* s) {
+  return dynamic_cast<rgw::io::Accounter*>(s->cio);
+}
 
-  uint64_t get_bytes_sent() { return bytes_sent; }
-  uint64_t get_bytes_received() { return bytes_received; }
-}; /* RGWStreamIO */
+static inline RGWRestfulIO* AWS_AUTHv4_IO(struct req_state* s) {
+  return static_cast<RGWRestfulIO*>(s->cio);
```

Comment:
i'm curious why the upcast from `rgw::io::BasicClient` to `RGWRestfulIO` is allowed here without `dynamic_cast`

and i suppose the same goes for `RESTFUL_IO` - how do we guarantee that the `BasicClient *cio` is actually a `RestfulClient` and not some other derived class?